kernel/src/log_segment.rs

```diff
@@ -112,13 +126,234 @@ impl LogSegment {
+            }
+            _ => {
+                let (commit_files, checkpoint_files, _) =
+                    Self::list_log_files_from_version(fs_client, &log_root, None)?;
+
+                (commit_files, checkpoint_files)
+            }
+        };
+
+        if omit_checkpoint_files {
+            checkpoint_files.clear();
+        }
+
+        // Commit file versions must satisfy the following:
+        // - Be greater than or equal to the start version
+        // - Be greater than the most recent checkpoint version if it exists
+        // - Be less than or equal to the end version.
+        if let Some(start_version) = start_version {
+            commit_files.retain(|log_path| log_path.version >= start_version);
+        }
+        if let Some(checkpoint_file) = checkpoint_files.first() {
+            commit_files.retain(|log_path| checkpoint_file.version < log_path.version);
+        }
+        if let Some(end_version) = end_version {
+            commit_files.retain(|log_path| log_path.version <= end_version);
+        }
+
+        // get the effective version from chosen files
+        let version_eff = commit_files
+            .last()
+            .or(checkpoint_files.first())
+            .ok_or(Error::MissingVersion)? // TODO: A more descriptive error
+            .version;
+        if let Some(end_version) = end_version {
+            require!(
+                version_eff == end_version,
+                Error::MissingVersion // TODO more descriptive error
+            );
+        }
+
+        // We assume listing returned ordered. If `in_order_commit_files` is false, we want reverse order.
+        if reversed_commit_files {
+            commit_files.reverse();
+        }
+
+        Ok(LogSegment {
+            end_version: version_eff,
+            log_root,
+            commit_files,
+            checkpoint_files,
+        })
+    }
+    pub(crate) fn list_log_files_from_version(
+        fs_client: &dyn FileSystemClient,
+        log_root: &Url,
+        version: Option<Version>,
+    ) -> DeltaResult<(Vec<ParsedLogPath>, Vec<ParsedLogPath>, i64)> {
+        let begin_version = version.unwrap_or(0);
+        let version_prefix = format!("{:020}", begin_version);
+        let start_from = log_root.join(&version_prefix)?;
+
+        let mut max_checkpoint_version = version.map_or(-1, |x| x as i64);
+        let mut checkpoint_files = vec![];
+        // We expect 10 commit files per checkpoint, so start with that size. We could adjust this based
+        // on config at some point
+        let mut commit_files = Vec::with_capacity(10);
+
+        for meta_res in fs_client.list_from(&start_from)? {
+            let meta = meta_res?;
+            // TODO this filters out .crc files etc which start with "." - how do we want to use these kind of files?
+            if let Some(parsed_path) = ParsedLogPath::try_from(meta)? {
+                if parsed_path.is_commit() {
+                    commit_files.push(parsed_path);
```

Comment:
future work: peek then push to validate ordering/gaps?

Reply:
I like that 👌